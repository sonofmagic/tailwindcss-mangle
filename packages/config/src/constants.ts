export const configName = 'tailwindcss-mangle'
