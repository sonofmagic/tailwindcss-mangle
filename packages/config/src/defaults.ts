import { defaultMangleClassFilter } from '@tailwindcss-mangle/shared'
import type { PatchUserConfig, UserConfig, MangleUserConfig } from './types'

export function getDefaultPatchConfig(): PatchUserConfig {
  return {
    output: {
      filename: '.tw-patch/tw-class-list.json',
      removeUniversalSelector: true,
      loose: true
    },
    tailwindcss: {}
  }
}

export function getDefaultMangleUserConfig(): MangleUserConfig {
  return {
    mangleClassFilter: defaultMangleClassFilter,
    include: ['**/*.{js,jsx,ts,tsx,svelte,vue}'],
    exclude: ['node_modules/**/*', '**/*.{css,scss,less,sass,postcss,html,htm}'],
    disabled: process.env.NODE_ENV === 'development',
    classListPath: '.tw-patch/tw-class-list.json',
    classMapOutput: {
      enable: false,
      filename: '.tw-patch/tw-map-list.json',
      loose: true
    }
  }
}

export function getDefaultUserConfig(): UserConfig {
  return {
    patch: getDefaultPatchConfig(),
    mangle: getDefaultMangleUserConfig()
  }
}
