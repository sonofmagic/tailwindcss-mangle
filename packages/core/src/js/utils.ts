import type babel from '@babel/core'
// const t = babel.types
export function getStringLiteralCalleeName(path: babel.NodePath<babel.types.StringLiteral>) {
  if (path.parentPath.isCallExpression()) {
    const callee = path.parentPath.get('callee')
    if (callee.isIdentifier()) {
      return callee.node.name
    }
  }
}

export function getTemplateElementCalleeName(path: babel.NodePath<babel.types.TemplateElement>) {
  if (path.parentPath.isTemplateLiteral()) {
    const pp = path.parentPath
    if (pp.parentPath.isCallExpression()) {
      const callee = pp.parentPath.get('callee')
      if (callee.isIdentifier()) {
        return callee.node.name
      }
    }
  }
}
