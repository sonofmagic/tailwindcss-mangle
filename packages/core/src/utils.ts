import micromatch from 'micromatch'

const { isMatch } = micromatch

export function escapeStringRegexp(str: string) {
  if (typeof str !== 'string') {
    throw new TypeError('Expected a string')
  }
  return str.replaceAll(/[$()*+.?[\\\]^{|}]/g, '\\$&').replaceAll('-', '\\x2d')
}

export function createGlobMatcher(pattern: string | string[] | undefined, fallbackValue: boolean = false) {
  if (pattern === undefined) {
    return function () {
      return fallbackValue
    }
  }
  return function (file: string) {
    return isMatch(file, pattern)
  }
}

export { defaultMangleClassFilter, isMap, isRegexp } from '@tailwindcss-mangle/shared'
