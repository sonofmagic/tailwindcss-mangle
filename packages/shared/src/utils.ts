export const preserveClassNames = [
  // https://tailwindcss.com/docs/transition-timing-function start
  // https://github.com/sonofmagic/tailwindcss-mangle/issues/21
  'ease-out',
  'ease-linear',
  'ease-in',
  'ease-in-out'
  // https://tailwindcss.com/docs/transition-timing-function end
]

// eslint-disable-next-line unicorn/no-array-reduce
const preserveClassNamesMap = preserveClassNames.reduce<Record<(typeof preserveClassNames)[number], true>>((acc, cur) => {
  acc[cur] = true
  return acc
}, {})

export const defaultMangleClassFilter = (className: string) => {
  if (preserveClassNamesMap[className]) {
    return false
  }
  // ignore className like 'filter','container'
  // it may be dangerous to mangle/rename all StringLiteral , so use /-/ test for only those with /-/ like:
  // bg-[#123456] w-1 etc...
  return /[:-]/.test(className)
}

export function groupBy<T>(arr: T[], cb: (arg: T) => string): Record<string, T[]> {
  if (!Array.isArray(arr)) {
    throw new TypeError('expected an array for first argument')
  }

  if (typeof cb !== 'function') {
    throw new TypeError('expected a function for second argument')
  }

  const result: Record<string, T[]> = {}
  for (const item of arr) {
    const bucketCategory = cb(item)
    const bucket = result[bucketCategory]

    if (Array.isArray(bucket)) {
      result[bucketCategory].push(item)
    } else {
      result[bucketCategory] = [item]
    }
  }

  return result
}

export const acceptChars = [...'abcdefghijklmnopqrstuvwxyz']

export function stripEscapeSequence(words: string) {
  return words.replaceAll('\\', '')
}

export function isRegexp(value: unknown) {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

export function isMap(value: unknown) {
  return Object.prototype.toString.call(value) === '[object Map]'
}

export function regExpTest(arr: (string | RegExp)[] = [], str: string) {
  if (Array.isArray(arr)) {
    for (const item of arr) {
      if (typeof item === 'string') {
        if (item === str) {
          return true
        }
      } else if (isRegexp(item)) {
        item.lastIndex = 0
        if (item.test(str)) {
          return true
        }
      }
    }
    return false
  }
  throw new TypeError("paramater 'arr' should be a Array of Regexp | String !")
}
