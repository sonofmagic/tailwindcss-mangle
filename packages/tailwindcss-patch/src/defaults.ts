import type { DeepRequired, PatchOptions } from './types'

export function getDefaultPatchOptions(): DeepRequired<PatchOptions> {
  return {
    overwrite: true,
  }
}
