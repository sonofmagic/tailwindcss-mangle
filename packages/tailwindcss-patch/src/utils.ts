import fss from 'node:fs'
import fs from 'node:fs/promises'
import type { SyncOpts } from 'resolve'
import pkg from 'resolve'

const { sync } = pkg

export function ensureFileContent(filepaths: string | string[]) {
  if (typeof filepaths === 'string') {
    filepaths = [filepaths]
  }
  let content
  for (const filepath of filepaths) {
    if (fss.existsSync(filepath)) {
      content = fss.readFileSync(filepath, {
        encoding: 'utf8',
      })
      break
    }
  }
  return content
}

export function requireResolve(id: string, opts?: SyncOpts) {
  return sync(id, opts)
}

export async function ensureDir(p: string) {
  try {
    await fs.access(p)
  }
  catch {
    await fs.mkdir(p, {
      recursive: true,
    })
  }
}
