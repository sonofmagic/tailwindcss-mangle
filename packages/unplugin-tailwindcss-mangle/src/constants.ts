export const pluginName = 'unplugin-tailwindcss-mangle'
