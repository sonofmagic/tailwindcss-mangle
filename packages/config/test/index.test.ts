import { resolve } from 'node:path'
import { existsSync } from 'node:fs'
import { deleteAsync } from 'del'
import { fixturesRoot } from './utils'
import { initConfig, getConfig } from '@/index'
import { getDefaultUserConfig } from '@/defaults'
import { configName } from '@/constants'

describe('config', () => {
  it('0.default', async () => {
    const cwd = resolve(fixturesRoot, './config/0.default')
    const configPath = resolve(cwd, `${configName}.config.ts`)
    if (existsSync(configPath)) {
      await deleteAsync(configPath)
    }

    await initConfig(cwd)
    expect(existsSync(configPath)).toBe(true)

    const { config } = await getConfig(cwd)
    expect(config).toEqual(getDefaultUserConfig())
  })

  it('1.change-options', async () => {
    const cwd = resolve(fixturesRoot, './config/1.change-options')
    const { config } = await getConfig(cwd)
    expect(config).toEqual({
      patch: {
        output: {
          filename: 'xxx/yyy/zzz.json',
          loose: false,
          removeUniversalSelector: false
        },
        tailwindcss: {
          cwd: 'aaa/bbb/cc'
        }
      }
    })
  })
})
