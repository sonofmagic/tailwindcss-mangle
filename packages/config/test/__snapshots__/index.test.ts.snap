// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > 2.mangle-options 1`] = `
{
  "mangle": {
    "classGenerator": {
      "log": true,
    },
    "classListPath": "zzzzz.json",
    "classMapOutput": {
      "enable": true,
      "filename": "ffff.json",
      "loose": false,
    },
    "disabled": false,
    "exclude": [
      "node_modules/**/*",
      "**/*.{css,scss,less,sass,postcss,html,htm}",
    ],
    "include": [
      "**/*.{js,jsx,ts,tsx,svelte,vue}",
    ],
    "mangleClassFilter": [Function],
  },
  "patch": {
    "output": {
      "filename": "xxx/yyy/zzz.json",
      "loose": false,
      "removeUniversalSelector": false,
    },
    "tailwindcss": {
      "cwd": "aaa/bbb/cc",
    },
  },
}
`;
